Add support for binding the depth buffer as a color target. Fixes Kurohyo depth sorting

hrydgard opened the pull request:
Fixes #9576 

Kurohyo uses 2d backdrops with 3d characters rendered on top of them, and to fill in the Z buffer to make the characters sort against the background correctly, it writes a depth image directly to the Z buffer.

The PSP doesn't have any special mode for this, instead this game uses 565 color mode, binds the Z buffer as a color buffer, and writes the image as a color image.

I am not introducing a game specific compat flag, hoping for the best, heh (that it won't break something else).

Since such rebinding is entirely impossible on PC, but we can write explicitly to depth, we find the corresponding framebuffer that last touched that depth buffer, bind it normally for rendering, and we use a new rendering mode where:

* Color writes are masked away (don't happen)
* Depth writes are enabled in ALWAYS mode (no testing, just writing)
* The fragment color value is interpreted as a 565 value overlapping a 16-bit depth value, and the computed depth value is written to the depth buffer

Result - the characters should clip correctly!

~~WIP, only works in Vulkan in straight-through depth mode so far, but the other backends will be fixed too.~~

hrydgard committed:
Add logging for some non-traditional uses of framebuffers.

hrydgard committed:
Implement color-to-depth for Vulkan, start implementing for D3D11

hrydgard committed:
D3D11 fix

hrydgard committed:
OpenGL fix

hrydgard committed:
D3D9 fix

hrydgard commented:
Hm, still has problems on Vulkan on Adreno, works on OpenGL though.

dante3732 commented:
Fix need enable or it work by default?

dante3732 commented:
It seems to work fine, I will write if there are problems.

dante3732 commented:
The character is partially visible through the car.
![1](https://user-images.githubusercontent.com/66036249/182139223-04486ad2-fd38-4e24-bf89-75634bcde49f.png)
![Uploading 2.png…]()

dante3732 commented:
Same
![Снимок экрана (2306)](https://user-images.githubusercontent.com/66036249/182141478-4753ef46-3292-4042-b79a-95657b68b154.png)

hrydgard commented:
Ah interesting, think we might be mistakenly filtering the depth texture. I'll get a fix in soon.

hrydgard committed:
Disable texture filtering when drawing color into depth

hrydgard commented:
I believe that change should take care of it.

hrydgard committed:
Avoid upscaling textures used for color-to-depth.

(Not a perfect method, it could theoretically have been upscaled in a
normal render pass...)

dante3732 commented:
> I believe that change should take care of it.

![Uploading 12.png…]()

hrydgard commented:
Also, does it look like that in all backends (Vulkan, D3D etc?)

dante3732 commented:
> Okay, I guess not, but it kinda has to be that, or something very related... Can I have a GE dump or savestate file from that location?

How make GE dump?

hrydgard commented:
https://github.com/hrydgard/ppsspp/wiki/How-to-create-a-frame-dump

dante3732 commented:
> Also, does it look like that in all backends (Vulkan, D3D etc?)

Only on D3D9 all fine. 
![324](https://user-images.githubusercontent.com/66036249/182156528-4765f507-1922-41a1-b2f3-5a411f10aaf8.png)

dante3732 commented:
> Okay, I guess not, but it kinda has to be that, or something very related... Can I have a GE dump or savestate file from that location?

Savestate https://drive.google.com/file/d/1-o47utQjASHd0EDx3EOaSkEfmXU_y5IU/view?usp=sharing

dante3732 commented:
Ge Dump https://drive.google.com/file/d/10aHhuhEuLsi8i_8nlzidwocI7zlO9Dlh/view?usp=sharing

hrydgard commented:
thanks!

hrydgard commented:
Yeah there's something weird going on:

source image:

![image](https://user-images.githubusercontent.com/130929/182158974-366e16ce-82de-4d61-8106-54c35b5fffbe.png)

resulting z buffer:

![image](https://user-images.githubusercontent.com/130929/182159075-04da96a0-1355-4542-a9c0-5e765b48f440.png)

Then it gets fuzzy because of some blur effect later in the frame, but the depth buffer is wrong right from the beginning. hmm

hrydgard committed:
Fix for D3D11 and GL, but not Vulkan for some weird reason

hrydgard committed:
Fix Vulkan as well

hrydgard commented:
Alright, I think I got it sorted now.
